tensorflow/contrib/learn/python/learn/datasets/mnist.py

```diff
@@ -168,22 +168,40 @@ def next_batch(self, batch_size, fake_data=False):
       return [fake_image for _ in xrange(batch_size)], [
           fake_label for _ in xrange(batch_size)
       ]
+
+    """ Get next batch """ 
     start = self._index_in_epoch
-    self._index_in_epoch += batch_size
-    if self._index_in_epoch > self._num_examples:
+    # Shuffle for the first epoch
+    if self._epochs_completed ==0 and start == 0 and shuffle:
```

Comment:
This changes the semantics by shuffling before the first epoch. Is that intentional? Is it necessary?